File: Data Structures in C/Linked Lists/Circular Linked List/CircularLinkedListDeletion.c
/*
   This program shows how to delete a node at the begining and at the end of a Circular Linked List using C
   Author - Rajarshi Sengupta
   Github - https://github.com/rajarshisg
   Date - 19/08/2020 (dd/mm/yyyy)
*/
#include<stdio.h>
#include<stdlib.h>
typedef struct Node
{
    int data;
    struct Node *next;
}Node;
Node *deleteBegining(Node *head)
{
    if(head==NULL)
    return NULL;
     Node *tail=head;
    while(tail->next!=head)
    {
       tail=tail->next;
    }
    Node *n=head;
    tail->next=n->next;
    head=n->next;
    free(n);
   return head;
}
Node *deleteEnd(Node *head)
{
    if(head==NULL)
    return NULL;
    Node *curr=head;
    Node *tail=head;
    while(curr->next->next!=head)
    {
       curr=curr->next;
       tail=curr->next;
    }
    curr->next=head;
    free(tail);
    return head;
}
void traverse(Node *head)
{
     Node *curr=head;
    do
    {
        printf("%d ",curr->data);
        curr=curr->next;
    } while(curr!=head);
    printf("\n");
}
int main()
{
    int num;
    printf("Enter the number of nodes :\n");
    scanf("%d",&num);
    Node *head=NULL;
    Node *n=NULL;
    Node *m=NULL;
    printf("Enter the elements :\n");
    for(int i=1;i<=num;i++)
    {
        if(head==NULL)
        {
            n=(Node *)malloc(sizeof(Node));
            scanf("%d",&n->data);
            n->next=NULL;
            head=n;
        }
        else
        {
            m=(Node *)malloc(sizeof(Node));
            scanf("%d",&m->data);
            m->next=NULL;
            n->next=m;
            n=m;
        }
        if(i==num)
        n->next=head;
    }
    printf("The Linked List is :\n");
    traverse(head);
    printf("Deleteing node at first position...\n");
    head=deleteBegining(head);
    traverse(head);
    printf("Deleteing node at last position...\n");
    head=deleteEnd(head);
    traverse(head);
    return 0;
}

File: Data Structures in JAVA/Linked Lists/Circular Linked Lists/src/CircularLinkedListDeletion.java
/*
   This program shows how insert a node at the begining and end in a Circular Linked List in JAVA
   Author - Rajarshi Sengupta
   Github - https://github.com/rajarshisg
   Date - 20/08/2020 (dd/mm/yyyy)
   
   
   Node class definition :
   public class Node<T> {
   T data;
   Node next;
   Node(T data){
	   this.data=data;
	   this.next=null;
   }
   
}

*/
import java.util.Scanner;
public class CircularLinkedListDeletion {
	Node deleteBegining(Node<Integer> head) {
		Node<Integer> tail=head;
	    while(tail.next!=head)
	    	tail=tail.next;
		tail.next=head.next;
		Node<Integer> n=head;
		n=null;
		head=tail.next;
		return head;
	}
	Node deleteEnd(Node<Integer> head) {
		Node<Integer> tail=head;
	    while(tail.next.next!=head)
	    	tail=tail.next;
	    Node<Integer> n=tail.next;
		tail.next=head;
		n=null;
		return head;
	}
	void traverse(Node<Integer> head) {
		 System.out.println("The Linked List is :");
		   Node curr=head;
		   do
		   {
			   System.out.print(curr.data+" ");
			   curr=curr.next;
		   }while(curr!=head);
		   System.out.println();
	}
   public static void main(String args[])
   {
	   Scanner sc=new Scanner(System.in);
	   Node<Integer> head=null,n=null,m=null;
	   System.out.println("Enter the number of nodes :");
	   int num=sc.nextInt();
	   System.out.println("Enter the elements :");
	   for(int i=1;i<=num;i++)
	   {
		   if(head==null)
		   {
			   n=new Node(sc.nextInt());
			   head=n;
		   }
		   else
		   {
			   m=new Node(sc.nextInt());
			   n.next=m;
			   n=n.next;
		   }
	   }
	   n.next=head; //last node points to first
	   CircularLinkedListDeletion obj=new CircularLinkedListDeletion();
	   obj.traverse(head);
	   System.out.println("Deleting the first node...");
	   head=obj.deleteBegining(head);
	   obj.traverse(head);
	   System.out.println("Deleting the last node...");
	   head=obj.deleteEnd(head);
	   obj.traverse(head);
   }
}
